XiEditor/EditViewController.swift

```diff
@@ -133,11 +133,10 @@ class EditViewController: NSViewController, EditViewDataSource, FindDelegate, Sc
     }
     
     /// Current language used for syntax highlighting
-    var currentLanguage: String? {
-        didSet {
-            updateLanguageMenu()
-        }
-    }
+    private var currentLanguage: String = ""
+
+    /// Object provides available and recently used languages
+    var languagesProvider: LanguagesProvider!
```

Comment:
Okay, so I've been thinking about this a bit:

I think it's important that we have a clear sense of what the source of truth is for language information. I'm thinking of the situation where (theoretically) a new plugin is loaded in core that adds a new language: What happens in this case? As currently implemented, the `LanguagesProvider` in the `AppDelegate` would get updated, but the individual views wouldn't end up being notified?

There are a variety of approaches to this. One simple one would be to have all access to the `languagesProvider` be explicitly shared; the view controller would be returning a reference to the document, which would be returning the shared object in the AppDelegate. Then when the state changed, we would iterate through the view controllers and notify them of the state change, and they can reload menus etc as needed.

Alternatively, when the state gets set in core we could go and set the new `languageProvider` on each view controller, and then handle any updates in a `didSet` method.

Another confusion here is that because `LanguagesStore` is a class, everyone should currently be sharing a single instance, and so the state _should_ be in sync; but `LanguagesProvider` isn't declared as a class protocol, so if it were implemented as a struct this would no longer be true?

Another thing I might think about is whether this state belongs more in the document or in the view controller. I think it might be reasonable to have it only be in the document? I could go either way, though.

Reply:
@cmy Thank you for review! 

You are right, class protocol constraint definitely needs to be added to `LanguagesProvider` 👍🏻 

Ideally `ViewController` should not contain any state. ViewController is an entity of UI / Presentation layer and its main responsibility to be a glue between model (view-model) and view and handle view's lifecycle. I definitely would not keep state in view controller and most likely it is Document object's responsibility.

I have been thinking about it a bit and I came up with possible solution for the problem. The idea is to have `LanguagesStore` as single source of truth and listeners which can listen to notifications about the updates. The `LanguageStore` itself can be updated by receiving notification from the core. I have created high level UML diagram to make it a little bit clear. Please, have a look:

![languages store-2](https://user-images.githubusercontent.com/1524024/47388232-f636c300-d711-11e8-8895-d469f048a10d.png)

Diagram is high level in some cases (Document and XiDocumentController, Core), but still brings visibility how data flows from presentation layer to core and to `LanguagesStore`.

What do you think?

Reply:
So I think we're broadly on the same page; what we want is a single source of truth and a notification mechanism.

We have options in terms of what our notification mechanism is. The simplest is that when the state changes, we iterate through our documents and call some `stateChanged()` function; then each document goes and checks the new state and performs some update logic. Another reasonable option is to lean on the Cocoa `NotificationCenter` infrastructure, which you're doing here. In _this_ case, though, I think it makes more sense to pass some reference type containing the state along as the notification's object; so in this case I would register for, say, the 'availableLanguagesChanged' notification, and the notification would contain a reference to some type, say, `CoreState`, (which can be a protocol) which has a member `var availableLanguages { get }`.

I think this is mostly what you're suggesting? The main thing I want to push is that we'd like to balance the benefits (clear separation of responsibilities, clear flow of data) against the costs of increasing architectural complexity; so whatever we do here should be something that we anticipate being the general model for communicating changes to shared state.

Reply:
@cmyr Yes, its true! This is what I am suggesting. Please, take a note that `LanguageStoringNotification` is actually an interface. Interface name is misleading and will be changed. Having interface makes a loose coupling and allows to do checking during compile time, while `NSNotification.Name` is actually a `String` and it is easy to break an implementation. So, I would not use `NotificationCenter`.

So, the basic idea is very simple: 

1. Subscriber implements protocol and be subscribed for receiving updates from `LanguagesStore`
2. `LanguagesStore` receives updates from the `Core`, updates it's state and notifies all subscribers by iterating over the collection and calling appropriate function
3. As you mentioned above, updated state of the store (as a type of lets say `LanguageStoringOutput`) can passed to the function as parameter

What do you think? Does it make sense?

Architecture is what allows you to manage complexity and makes it easy to adapt the code base for new requirements. So, there is always some level of architecture complexity. But, I agree, there should be a trade off.

Please, let know what do you think about my suggestion. Thank you!

Reply:
Okay, thanks for the explanation. I think we're on the same page, and this approach sounds reasonable. 👍

Reply:
My one thought is that we might want to eventually expand this so that we have a single mechanism for various types of state (themes are an obvious example), so I'd keep that in mind; for instance we might want multiple types of events that can be subscribed to, _or_ we might want them to be a package deal.

Reply:
@cmyr Thank you very much for the feedback! Yes, I am also thinking about generic solution that can be used in variety of cases. I was busy at work these days and going to have look at it on the weekend. Thanks!